### PR TITLE
mgr/cephadm: cephadm bootstrap add --apply-spec <cluster.yaml>

### PR DESCRIPTION
**Pull request by Daniel-Pivonka**
Have a single command when setting up a cluster for Day 1

example spec

```
---
service_type: host
addr: jmo-node-00
hostname: jmo-node-00
---
service_type: host
addr: jmo-node-01
hostname: jmo-node-01
labels:
  - grafana
---
service_type: host
addr: jmo-node-02
hostname: jmo-node-02
---
service_type: grafana
placement:
  label: "grafana"
---
service_type: osd
placement:
  host_pattern: '*'
data_devices:
  all: true

```
Fixes: [https://tracker.ceph.com/issues/44873]

Signed-off-by: Daniel-Pivonka <dpivonka@redhat.com>

### CODE REVIEW

**Comment by Daniel-Pivonka** [comment]
Still working on this but I wanted to post this to get feedback while its in a POC working state

File: src/pybind/mgr/orchestrator/module.py [hunk 1]
```diff
@@ -844,11 +844,21 @@ def _apply_misc(self, service_type=None, placement=None, unmanaged=False, inbuf=
...
                 raise OrchestratorValidationError(usage)
             content: Iterator = yaml.load_all(inbuf)
-            specs = [ServiceSpec.from_json(s) for s in content]
+
+            specs = []
+            for s in content:
+                if 'service_type' in s and s['service_type'] == 'host':
+                    spec = HostSpec.from_json(s)
+                    spec.service_type = 'host'
+                    specs.insert(0, spec)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:858]: changing the order here is a bit strange.

File: src/pybind/mgr/orchestrator/module.py [hunk 2]
```diff
@@ -844,11 +844,21 @@ def _apply_misc(self, service_type=None, placement=None, unmanaged=False, inbuf=
...
             if service_type or placement or unmanaged:
                 raise OrchestratorValidationError(usage)
             content: Iterator = yaml.load_all(inbuf)
-            specs = [ServiceSpec.from_json(s) for s in content]
+
+            specs = []
+            for s in content:
+                if 'service_type' in s and s['service_type'] == 'host':
+                    spec = HostSpec.from_json(s)
+                    spec.service_type = 'host'
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:857]: `HostSpec` does not have a `service_type` attribute. One way would be to simply add it. 

I have the feeling that differentiating here between `HostSpec` and  `ServiceSpec` here will need to get refactored eventually.  Maybe we can move this code to `ServiceSpec.from_json()`?

File: src/pybind/mgr/orchestrator/module.py [hunk 3]
```diff
@@ -844,11 +844,21 @@ def _apply_misc(self, service_type=None, placement=None, unmanaged=False, inbuf=
   ceph orch apply -i <yaml spec>
   ceph orch apply <service_type> <placement> [--unmanaged]
         """
+        results = []
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:847]: no need to have this.

File: src/pybind/mgr/cephadm/module.py [hunk 4]
```diff
@@ -2759,7 +2759,15 @@ def add_mgr(self, spec):
         return self._add_daemon('mgr', spec, self._create_mgr)
 
     def _apply(self, spec: ServiceSpec) -> str:
-        if spec.placement.is_empty():
+
+        if spec.service_type == 'host':
+            completion = self.add_host(spec)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:2764]: This should fail badly in mypy. either we change the type hierarchy (i.e. move `HostSpec` and `ServiceSpec` into a common type hierarchy) or we differentiate between hosts and `ServiceSpec` somewhere else.

File: src/cephadm/cephadm [hunk 5]
```diff
@@ -2591,6 +2591,24 @@ def command_bootstrap():
...
+                    line = line.replace('\n', '')
+                    split = line.split(': ')
+                    if split[1] != host:
+                        logger.info('Adding ssh key to %s' % split[1])
+                        out, err, code = call_throws(['ssh-copy-id', '-f', '-i', '/etc/ceph/ceph.pub', 'root@%s' % split[1]])
+
+        mounts = {}
+        mounts[pathify(args.apply_spec)] = '/tmp/spec.yml:z'
+
+        out = cli(['orch', 'apply', '-i', '/tmp/spec.yml'], extra_mounts=mounts)
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:2610]: instead of extra mounts, what about stdin?

File: src/pybind/mgr/orchestrator/module.py (hunk 1 above)
Comment (Daniel-Pivonka) [reply src/pybind/mgr/orchestrator/module.py:857]: hosts need to be at the front of the list of specs so they are added to the cluster before other specs that rely on them are applied. Ill try to find a way make this cleaner though

File: src/cephadm/cephadm (hunk 5 above)
Comment (Daniel-Pivonka) [reply src/cephadm/cephadm:2610]: I give that a try this is how files were passed to the cli in other places though so i just went with that

**Comment by Daniel-Pivonka** [comment]
Ive made a HostSpec class that is a child class of ServiceSpec. 
Im still using the original HostSpec class further down the code path though. 
That changed the yaml file to this

```
---
service_type: host
spec:
    addr: jmo-node-00
    hostname: jmo-node-00
---
service_type: host
spec:
    addr: jmo-node-01
    hostname: jmo-node-01
    labels:
    - grafana
---
service_type: host
spec:
    addr: jmo-node-02
    hostname: jmo-node-02
---
service_type: grafana
placement:
  label: "grafana"
---
service_type: osd
placement:
  host_pattern: '*'
data_devices:
  all: true
```

File: src/python-common/ceph/deployment/service_spec.py [hunk 6]
```diff
@@ -605,3 +606,23 @@ def validate_add(self):
         if not self.pool:
             raise ServiceSpecValidationError(
                 'Cannot add ISCSI: No Pool specified')
+
+class HostSpec(ServiceSpec):
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/service_spec.py:610]: So, is the hostSpec actually a ServiceSpec, or do we need another base class?

I mean, `placement` doesn't make sense for HostSpec.

image

File: src/pybind/mgr/cephadm/module.py [hunk 7]
```diff
@@ -1690,6 +1690,15 @@ def add_host(self, spec):
...
         return "Added host '{}'".format(spec.hostname)
 
+    def _add_host(self, spec):
+        # type: (HostSpec) -> str
+        completion = self.add_host(spec)
+        self.process([completion])
+        while completion.needs_result:
+            time.sleep(1)
+        orchestrator.raise_if_exception(completion)
+        return completion.result_str()
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1700]: I'd do it the other way around. 

1. rename `def add_host(self, spec: HostSpec):` to _add_host
2. add

```python
@async_completion
def add_host(self, spec):
    return self._add_host(spec)
```

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
apply this and then it type checks:

```patch
diff --git a/src/pybind/mgr/cephadm/module.py b/src/pybind/mgr/cephadm/module.py
index 0d477fe82b..5a7b166291 100644
--- a/src/pybind/mgr/cephadm/module.py
+++ b/src/pybind/mgr/cephadm/module.py
@@ -27,6 +27,7 @@ from mgr_module import MgrModule, HandleCommandResult
 import orchestrator
 from orchestrator import OrchestratorError, OrchestratorValidationError, HostSpec, \
     CLICommandMeta
+from orchestrator._interface import GenericSpec
 
 from . import remotes
 from . import utils
@@ -1022,7 +1023,10 @@ class CephadmOrchestrator(orchestrator.Orchestrator, MgrModule):
         return list(self.inventory.filter_by_label(label))
 
     @trivial_completion
-    def add_host(self, spec):
+    def add_host(self, host_spec: HostSpec) -> str:
+        return self._add_host(host_spec)
+
+    def _add_host(self, spec):
         # type: (HostSpec) -> str
         """
         Add a host to be managed by the orchestrator.
@@ -1045,15 +1049,6 @@ class CephadmOrchestrator(orchestrator.Orchestrator, MgrModule):
         self.log.info('Added host %s' % spec.hostname)
         return "Added host '{}'".format(spec.hostname)
 
-    def _add_host(self, spec):
-        # type: (HostSpec) -> str
-        completion = self.add_host(spec)
-        self.process([completion])
-        while completion.needs_result:
-            time.sleep(1)
-        orchestrator.raise_if_exception(completion)
-        return completion.result_str()
-
     @trivial_completion
     def remove_host(self, host):
         # type: (str) -> str
@@ -1890,7 +1885,13 @@ class CephadmOrchestrator(orchestrator.Orchestrator, MgrModule):
         # type: (ServiceSpec) -> List[str]
         return self._add_daemon('mgr', spec, self.mgr_service.create)
 
-    def _apply(self, spec: ServiceSpec) -> str:
+    def _apply(self, spec: GenericSpec) -> str:
+        if spec.service_type == 'host':
+            return self._add_host(cast(HostSpec, spec))
+
+        return self._apply_service_spec(cast(ServiceSpec, spec))
+
+    def _apply_service_spec(self, spec: ServiceSpec) -> str:
         if spec.placement.is_empty():
             # fill in default placement
             defaults = {
@@ -1929,8 +1930,6 @@ class CephadmOrchestrator(orchestrator.Orchestrator, MgrModule):
     @trivial_completion
     def apply(self, specs: List[ServiceSpec], hosts: List[HostSpec]=[]):
         results = []
-        for host in hosts:
-            results.append(self._add_host(spec))
         for spec in specs:
             results.append(self._apply(spec))
         return results
diff --git a/src/pybind/mgr/orchestrator/_interface.py b/src/pybind/mgr/orchestrator/_interface.py
index 0ac903cd67..87c2561c77 100644
--- a/src/pybind/mgr/orchestrator/_interface.py
+++ b/src/pybind/mgr/orchestrator/_interface.py
@@ -855,24 +855,25 @@ class Orchestrator(object):
         """
         raise NotImplementedError()
 
-    def apply(self, specs: List[ServiceSpec], hosts=[]) -> Completion:
+    def apply(self, specs: Sequence["GenericSpec"]) -> Completion:
         """
         Applies any spec
         """
-        fns: Dict[str, Callable[[ServiceSpec], Completion]] = {
+        fns: Dict[str, function] = {
             'alertmanager': self.apply_alertmanager,
             'crash': self.apply_crash,
             'grafana': self.apply_grafana,
-            'iscsi': cast(Callable[[ServiceSpec], Completion], self.apply_iscsi),
+            'iscsi': self.apply_iscsi,
             'mds': self.apply_mds,
             'mgr': self.apply_mgr,
             'mon': self.apply_mon,
-            'nfs': cast(Callable[[ServiceSpec], Completion], self.apply_nfs),
+            'nfs': self.apply_nfs,
             'node-exporter': self.apply_node_exporter,
-            'osd': cast(Callable[[ServiceSpec], Completion], lambda dg: self.apply_drivegroups([dg])),
+            'osd': lambda dg: self.apply_drivegroups([dg]),
             'prometheus': self.apply_prometheus,
             'rbd-mirror': self.apply_rbd_mirror,
-            'rgw': cast(Callable[[ServiceSpec], Completion], self.apply_rgw),
+            'rgw': self.apply_rgw,
+            'host': self.add_host,
         }
 
         def merge(ls, r):
@@ -882,10 +883,12 @@ class Orchestrator(object):
 
         spec, *specs = specs
 
-        completion = fns[spec.service_type](spec)
+        fn = cast(Callable[[GenericSpec], Completion], fns[spec.service_type])
+        completion = fn(spec)
         for s in specs:
             def next(ls):
-                return fns[s.service_type](s).then(lambda r: merge(ls, r))
+                fn = cast(Callable[[GenericSpec], Completion], fns[spec.service_type])
+                return fn(s).then(lambda r: merge(ls, r))
             completion = completion.then(next)
         return completion
 
@@ -1163,6 +1166,7 @@ class HostSpec(object):
                  labels=None,  # type: Optional[List[str]]
                  status=None,  # type: Optional[str]
                  ):
+        self.service_type = 'host'
         #: the bare hostname on the host. Not the FQDN.
         self.hostname = hostname  # type: str
 
@@ -1206,7 +1210,11 @@ class HostSpec(object):
         return self.hostname == other.hostname and \
                self.addr == other.addr and \
                self.labels == other.labels
-       
+
+
+GenericSpec = Union[ServiceSpec, HostSpec]
+
+
 def from_json(spec):
     # type: (dict) -> Union[ServiceSpec, HostSpec]
     if 'service_type' in spec and spec['service_type'] == 'host':
diff --git a/src/pybind/mgr/orchestrator/module.py b/src/pybind/mgr/orchestrator/module.py
index 7cd8e34ac3..de16aa70e7 100644
--- a/src/pybind/mgr/orchestrator/module.py
+++ b/src/pybind/mgr/orchestrator/module.py
@@ -19,7 +19,7 @@ from ._interface import OrchestratorClientMixin, DeviceLightLoc, _cli_read_comma
     raise_if_exception, _cli_write_command, TrivialReadCompletion, OrchestratorError, \
     NoOrchestrator, OrchestratorValidationError, NFSServiceSpec, \
     RGWSpec, InventoryFilter, InventoryHost, HostSpec, CLICommandMeta, \
-    ServiceDescription, DaemonDescription, IscsiServiceSpec, from_json
+    ServiceDescription, DaemonDescription, IscsiServiceSpec, from_json, GenericSpec
 
 
 def nice_delta(now, t, suffix=''):
@@ -875,22 +875,14 @@ Usage:
                 raise OrchestratorValidationError(usage)
             content: Iterator = yaml.load_all(inbuf)
 
-            specs: List[ServiceSpec] = []
-            hosts: List[HostSpec] = []
-            for s in content:
-                spec = from_json(s)
-                if isinstance(spec, HostSpec):
-                    hosts.append(spec)
-                else:
-                    specs.append(spec)
-
-            completion = self.apply(specs, hosts)
+            specs: List[GenericSpec] = [from_json(s) for s in content]
 
         else:
             placmentspec = PlacementSpec.from_string(placement)
             assert service_type
             specs = [ServiceSpec(service_type, placement=placmentspec, unmanaged=unmanaged)]
-            completion = self.apply(specs)
+
+        completion = self.apply(specs)
 
         self._orchestrator_wait([completion])
         raise_if_exception(completion)
diff --git a/src/seastar b/src/seastar
index 0523b0fac4..471048f88e 160000
```

File: src/pybind/mgr/orchestrator/_interface.py [hunk 8]
```diff
@@ -855,24 +912,25 @@ def list_daemons(self, service_name=None, daemon_type=None, daemon_id=None, host
         """
         raise NotImplementedError()
 
-    def apply(self, specs: List[ServiceSpec]) -> Completion:
+    def apply(self, specs: List[GenericSpec]) -> Completion:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:915]: you can forward-declare things by using strings:

```suggestion
    def apply(self, specs: List["GenericSpec"]) -> Completion:
```

this way you don't need to move HostSpec

File: src/pybind/mgr/orchestrator/_interface.py [hunk 9]
```diff
@@ -1151,55 +1211,13 @@ def upgrade_available(self):
...
-        return "<HostSpec>({})".format(', '.join(map(repr, args)))
-
-    def __eq__(self, other):
-        # Let's omit `status` for the moment, as it is still the very same host.
-        return self.hostname == other.hostname and \
-               self.addr == other.addr and \
-               self.labels == other.labels
-
+       
+def from_json(spec):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:1215]: `from_json` is a bit too generic for my taste. Maybe we can find a better name? 

* `generic_spec_from_json`
* `json_to_generic_spec`
* `json_to_host_or_service_spec`
* `json_to_spec`

??

Comment (Daniel-Pivonka) [reply src/pybind/mgr/orchestrator/_interface.py]: went with json_to_generic_spec

**Comment by sebastian-philipp** [comment]
I think this is related to your code????

http://pulpito.ceph.com/swagner-2020-05-20_10:26:42-rados-wip-swagner-testing-2020-05-20-0959-distro-basic-smithi/5072690

```
2020-05-20T10:52:32.573 INFO:teuthology.orchestra.run.smithi149:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph --log-early orch set backend test_orchestrator
2020-05-20T10:52:32.885 INFO:teuthology.orchestra.run.smithi149:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph orch apply osd -i -
2020-05-20T10:52:33.184 INFO:tasks.ceph.mgr.x.smithi149.stderr:2020-05-20T10:52:33.181+0000 7f9ce6542700 -1 mgr handle_command module 'orchestrator' command handler threw exception: expected string or bytes-like object
2020-05-20T10:52:33.185 INFO:tasks.ceph.mgr.x.smithi149.stderr:2020-05-20T10:52:33.182+0000 7f9ce6542700 -1 mgr.server reply reply (22) Invalid argument Traceback (most recent call last):
2020-05-20T10:52:33.185 INFO:tasks.ceph.mgr.x.smithi149.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1171, in _handle_command
2020-05-20T10:52:33.185 INFO:tasks.ceph.mgr.x.smithi149.stderr:    return self.handle_command(inbuf, cmd)
2020-05-20T10:52:33.186 INFO:tasks.ceph.mgr.x.smithi149.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 113, in handle_command
2020-05-20T10:52:33.186 INFO:tasks.ceph.mgr.x.smithi149.stderr:    return dispatch[cmd['prefix']].call(self, cmd, inbuf)
2020-05-20T10:52:33.186 INFO:tasks.ceph.mgr.x.smithi149.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 311, in call
2020-05-20T10:52:33.186 INFO:tasks.ceph.mgr.x.smithi149.stderr:    return self.func(mgr, **kwargs)
2020-05-20T10:52:33.186 INFO:tasks.ceph.mgr.x.smithi149.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 75, in <lambda>
2020-05-20T10:52:33.187 INFO:tasks.ceph.mgr.x.smithi149.stderr:    wrapper_copy = lambda *l_args, **l_kwargs: wrapper(*l_args, **l_kwargs)
2020-05-20T10:52:33.187 INFO:tasks.ceph.mgr.x.smithi149.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 66, in wrapper
2020-05-20T10:52:33.187 INFO:tasks.ceph.mgr.x.smithi149.stderr:    return func(*args, **kwargs)
2020-05-20T10:52:33.187 INFO:tasks.ceph.mgr.x.smithi149.stderr:  File "/usr/share/ceph/mgr/orchestrator/module.py", line 666, in _apply_osd
2020-05-20T10:52:33.187 INFO:tasks.ceph.mgr.x.smithi149.stderr:    raise_if_exception(completion)
2020-05-20T10:52:33.187 INFO:tasks.ceph.mgr.x.smithi149.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 633, in raise_if_exception
2020-05-20T10:52:33.188 INFO:tasks.ceph.mgr.x.smithi149.stderr:    raise e
2020-05-20T10:52:33.188 INFO:tasks.ceph.mgr.x.smithi149.stderr:TypeError: expected string or bytes-like object
2020-05-20T10:52:33.188 INFO:tasks.ceph.mgr.x.smithi149.stderr:
```

**Comment by sebastian-philipp** [comment]
Do you also want to amend the man page? https://docs.ceph.com/docs/master/man/8/cephadm/

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
QA passed (more or less)

http://pulpito.front.sepia.ceph.com/swagner-2020-05-27_13:59:32-rados:cephadm-wip-swagner3-testing-2020-05-27-1219-distro-basic-smithi/

failures due to:

* #35273
* #34860

**Comment by Daniel-Pivonka** [comment]
should be all set. I added the support for provided ssh keys from #35195

**Comment by sebastian-philipp** [comment]
success: http://pulpito.ceph.com/swagner-2020-06-02_09:42:16-rados:cephadm-wip-swagner-testing-2020-06-02-0940-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
not that this is incomplete: `cephadm boostrap --apply-spec` make cephadm unresposive, when adding hosts. Thus we have to make the remote calls asynchronous in cephadm.

**Pull request merged**